extensions/trackedfile/src/main/java/io/deephaven/extensions/trackedfile/CompletableLocalOutputStream.java

```diff
@@ -0,0 +1,232 @@
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.channels.Channels;
+
+/**
+ * A {@link CompletableOutputStream} that writes to a temporary shadow file paths in the same directory to prevent
+ * overwriting any existing data in case of failure.
+ */
+public class CompletableLocalOutputStream extends CompletableOutputStream {
+
+    private final File firstCreatedDir;
+    private final File destFile;
+    private final File shadowDestFile;
+    private final OutputStream shadowDelegateStream; // Writes to the shadow file
+
+    private boolean done;
+    private boolean closed;
+    private boolean installedShadowFiles;
+
+    private static final Logger log = LoggerFactory.getLogger(CompletableLocalOutputStream.class);
+
+    CompletableLocalOutputStream(
+            @NotNull final File destFile,
+            @NotNull final TrackedSeekableChannelsProvider provider,
+            final int bufferSizeHint) throws IOException {
+        this.firstCreatedDir = prepareDestinationFileLocation(destFile);
+        this.destFile = destFile;
+        deleteBackupFile(destFile);
+        this.shadowDestFile = getShadowFile(destFile);
+        this.shadowDelegateStream = new BufferedOutputStream(Channels.newOutputStream(
+                provider.getWriteChannel(shadowDestFile)), bufferSizeHint);
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        verifyNotClosed();
+        shadowDelegateStream.write(b);
+    }
+
+    @Override
+    public void write(byte[] b) throws IOException {
+        verifyNotClosed();
+        shadowDelegateStream.write(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        verifyNotClosed();
+        shadowDelegateStream.write(b, off, len);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        verifyNotClosed();
+        shadowDelegateStream.flush();
+    }
+
+    public void done() {
+        done = true;
+    }
+
+    public void complete() throws IOException {
+        done();
+        shadowDelegateStream.close();
+        installShadowFile(destFile, shadowDestFile);
+        installedShadowFiles = true;
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        shadowDelegateStream.close();
+        deleteBackupFileNoExcept(destFile);
+        closed = true;
+    }
+
+    public void rollback() {
+        if (installedShadowFiles) {
+            rollbackShadowFiles(destFile);
+        }
+        // noinspection ResultOfMethodCallIgnored
+        shadowDestFile.delete();
+        if (firstCreatedDir != null) {
+            log.error().append("Cleaning up potentially incomplete table destination path starting from ")
+                    .append(firstCreatedDir.getAbsolutePath()).endl();
+            FileUtils.deleteRecursivelyOnNFS(firstCreatedDir);
+        }
+    }
```

Comment:
Deferring review of this block of code until you implement the intended state machine.